File: JAVA/java/src/StepByStepDirectionsFromaBinaryTreeNodetoAnother.java
import java.util.*;

public class StepByStepDirectionsFromaBinaryTreeNodetoAnother {

    public String getDirections(TreeNode root, int startValue, int destValue) {
        // Find the Lowest Common Ancestor (LCA) of start and destination nodes
        TreeNode lowestCommonAncestor = findLowestCommonAncestor(root, startValue, destValue);

        StringBuilder pathToStart = new StringBuilder();
        StringBuilder pathToDest = new StringBuilder();

        // Find paths from LCA to start and destination nodes
        findPath(lowestCommonAncestor, startValue, pathToStart);
        findPath(lowestCommonAncestor, destValue, pathToDest);

        StringBuilder directions = new StringBuilder();

        // Add "U" for each step to go up from start to LCA
        directions.append("U".repeat(pathToStart.length()));

        // Append the path from LCA to destination
        directions.append(pathToDest);

        return directions.toString();
    }

    private TreeNode findLowestCommonAncestor(TreeNode node, int value1, int value2) {
        if (node == null) return null;

        if (node.val == value1 || node.val == value2) return node;

        TreeNode leftLCA = findLowestCommonAncestor(node.left, value1, value2);
        TreeNode rightLCA = findLowestCommonAncestor(node.right, value1, value2);

        if (leftLCA == null) return rightLCA;
        else if (rightLCA == null) return leftLCA;
        else return node; // Both values found, this is the LCA
    }

    private boolean findPath(TreeNode node, int targetValue, StringBuilder path) {
        if (node == null) return false;

        if (node.val == targetValue) return true;

        // Try left subtree
        path.append("L");
        if (findPath(node.left, targetValue, path)) {
            return true;
        }
        path.setLength(path.length() - 1); // Remove last character

        // Try right subtree
        path.append("R");
        if (findPath(node.right, targetValue, path)) {
            return true;
        }
        path.setLength(path.length() - 1); // Remove last character

        return false;
    }

    // Helper function to build a tree from a list
    public TreeNode buildTreeFromList(Integer[] values) {
        if (values.length == 0 || values[0] == null) return null;

        TreeNode root = new TreeNode(values[0]);
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        int i = 1;

        while (i < values.length) {
            TreeNode current = queue.poll();
            if (values[i] != null) {
                current.left = new TreeNode(values[i]);
                queue.add(current.left);
            }
            i++;
            if (i < values.length && values[i] != null) {
                current.right = new TreeNode(values[i]);
                queue.add(current.right);
            }
            i++;
        }
        return root;
    }

    public static void main(String[] args) {
        StepByStepDirectionsFromaBinaryTreeNodetoAnother solver = new StepByStepDirectionsFromaBinaryTreeNodetoAnother();
        Integer[] treeValues = {5, 1, 2, 3, null, 6, 4};
        int startValue = 3;
        int destValue = 6;

        TreeNode root = solver.buildTreeFromList(treeValues);
        String result = solver.getDirections(root, startValue, destValue);
        System.out.println(result); // Expected output: "UURL"
    }
}

File: python/Step-By-Step Directions From a Binary Tree Node to Another.py
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class StepByStepDirectionsFromaBinaryTreeNodetoAnother:
    def get_directions(self, root, start_value, dest_value):
        # Find the Lowest Common Ancestor (LCA) of start and destination nodes
        lowest_common_ancestor = self.find_lowest_common_ancestor(root, start_value, dest_value)

        path_to_start = []
        path_to_dest = []

        # Find paths from LCA to start and destination nodes
        self.find_path(lowest_common_ancestor, start_value, path_to_start)
        self.find_path(lowest_common_ancestor, dest_value, path_to_dest)

        directions = []

        # Add "U" for each step to go up from start to LCA
        directions.extend(['U'] * len(path_to_start))

        # Append the path from LCA to destination
        directions.extend(path_to_dest)

        return ''.join(directions)

    def find_lowest_common_ancestor(self, node, value1, value2):
        if not node:
            return None

        if node.val == value1 or node.val == value2:
            return node

        left_lca = self.find_lowest_common_ancestor(node.left, value1, value2)
        right_lca = self.find_lowest_common_ancestor(node.right, value1, value2)

        if left_lca and right_lca:
            return node

        return left_lca if left_lca else right_lca

    def find_path(self, node, target_value, path):
        if not node:
            return False

        if node.val == target_value:
            return True

        # Try left subtree
        path.append('L')
        if self.find_path(node.left, target_value, path):
            return True
        path.pop()

        # Try right subtree
        path.append('R')
        if self.find_path(node.right, target_value, path):
            return True
        path.pop()

        return False


def build_tree_from_list(values):
    if not values:
        return None

    root = TreeNode(values[0])
    queue = deque([root])
    i = 1
    while i < len(values):
        node = queue.popleft()
        if i < len(values) and values[i] is not None:
            node.left = TreeNode(values[i])
            queue.append(node.left)
        i += 1
        if i < len(values) and values[i] is not None:
            node.right = TreeNode(values[i])
            queue.append(node.right)
        i += 1
    return root


if __name__ == "__main__":
    # Example usage
    tree_values = [5, 1, 2, 3, None, 6, 4]
    start_value = 3
    dest_value = 6

    root = build_tree_from_list(tree_values)

    solver = StepByStepDirectionsFromaBinaryTreeNodetoAnother()
    result = solver.get_directions(root, start_value, dest_value)
    print(result)  # Expected output: "UURL"
